### DISCUSSION TITLE
如何将手写Tensor IR实现与Relax 算子对接？

### DISCUSSION THREAD

**Discussion by qelk123** [mlc-ai/mlc-zh#89] [Q&A]
在第六讲中，借助BlockBuilder构造Tensor IR，按我的理解是通过写call_function_map 手工指定torch算子与TE ComputeOP之间的关系进行转换，再通过create_prim_func实现到Tensor IR 的lowering的。请问，TE 这一层次的Compute声明是否可以省略呢？特别是对于某些用TE这套DSL难以描述的算子或者是我已给出的高效Tensor IR实现的情形，是否能够绕过TE这一层呢？

**Comment by Hzfengsy** [comment]
从原理上看`emit_te`等价于一下几步操作：
1. CreatePrimFunc from TE
2. Add PrimFunc to the IRModule
3. Create `call_tir` with the generated PrimFunc
从原理上来说绕过TE完全没有问题，从目前的实现来看我们还没有实现相关的接口

**Reply by qelk123** [reply]
感谢回答！
那目前是没有办法在部署网络的过程中直接用上用tvm.script手写的算子tensor IR实现是吗？（不包括用tvm.script手写类似于tensorize的schedule）目前主要的做法感觉还是在TE这一层借助tvm.tir.ir_builder写定制的tir算子实现？

**Comment by hbsun2113** [comment]
@qelk123 请问您是要省略tensor expression，直接使用TensorIR function吗？
